### PR TITLE
More granular options parsing for Environment

### PR DESCRIPTION
**Pull request by briadam**
I'm seeking preliminary feedback on this before forging ahead.

Tests break when Boost Program Options are enabled, but they seem to on dev as well.  Guidance?
  FAIL: test_BoostInputOptionsParser
  FAIL: t01_valid_cycle/rtest01.sh
  (HANGS or SLOW): test_mala

I'm going to mark up the code myself with questions.  Also these general ones:
 * Should I further condense code shared between FullEnvironment::construct(...) variants?
 * Do we anticipate a need for options_have_been_used checking as in GPMSA?

### CODE REVIEW

File: src/core/src/Environment.C
Comment (briadam) [inline src/core/src/Environment.C:233]: I'm now default constructing EnvOptionsValues if not passed in.  Is that a problem?  Should I remove checks such as:
  if (m_optionsObj == NULL) return ".";
  queso_require_msg(m_optionsObj, "m_optionsObj variable is NULL");

  What about for EmptyEnvironment?  Is it important that it have options = NULL as opposed to just default constructed options?  The only use case is for member BaseTKGroup::m_emptyEnv which doesn't appear to get used anywhere...

File: src/core/inc/EnvironmentOptions.h
Comment (briadam) [inline src/core/inc/EnvironmentOptions.h]: I kept pointers such as const char* and Env* for consistency with ctor instead of changing APIs to use std::string and Env&

File: src/core/src/EnvironmentOptions.C
Comment (briadam) [inline src/core/src/EnvironmentOptions.C:51]: Why do we default construct a BoostInputOptionsParser for the empty input file case?

File: src/core/src/EnvironmentOptions.C
Comment (briadam) [inline src/core/src/EnvironmentOptions.C:169]: Why aren't debug parameters parsed from the input file?  Should I add them?

File: src/core/src/EnvironmentOptions.C
Comment (briadam) [inline src/core/src/EnvironmentOptions.C]: I'm converting the set<unsigned int> back to string for default handling.  Is the trailing whitespace a problem?

File: src/core/src/EnvironmentOptions.C
Comment (briadam) [inline src/core/src/EnvironmentOptions.C]: Should parse be robust to the case of an empty input file string?  For now, I'm leaving the error in place.  Seems calling parse means you really want to parse, but I could also see just silently skipping the parse if nothing to do.

**Review by briadam — COMMENTED** [review]
Questions inlined for @dmcdougall  and @roystgnr ...

File: src/core/src/Environment.C
Comment (roystgnr) [reply src/core/src/Environment.C:233]: We're going with "file options override constructor-time C++ options, right?  In that case default-constructing EnvOptionsValues seems like the right thing to do; it gives us a copy of the default options, unless they've been overridden in C++, and either way they're ready for the "input files can override us still further" code path.

Comment (roystgnr) [reply src/core/src/Environment.C:233]: The ```return ".";``` does seem redundant now, since that's the default value for ```EnvOptionsValues::m_subDisplayFileName```.

Comment (roystgnr) [reply src/core/src/Environment.C:233]: I'd leave the queso_require in there.  Since it's not in user hands anymore we could turn it into a queso_assert if you like, but it's also not in an inner loop anymore so there's no compelling reason to.

Comment (roystgnr) [reply src/core/src/Environment.C:233]: And there seems no reason not to give EmptyEnvironment default options, to make it less useless.

File: src/core/src/Environment.C
Comment (roystgnr) [inline src/core/src/Environment.C:1264]: This is a semantic change, not just an improved capability, so let's be wary of regressions.  I can't imagine anyone was actually *using* the behavior of "pass an input file name and a non-null option object pointer in and expect the former to be ignored", but I've seen stranger code in the past.

File: src/core/inc/EnvironmentOptions.h
Comment (roystgnr) [reply src/core/inc/EnvironmentOptions.h]: I wouldn't worry about consistency in a brand-new API, and I wouldn't worry too much about stylistic consistency in QUESO.

Using a reference is IMHO always better than a pointer to a class when you expect the pointer to never be null.

Using a char pointer rather than converting to std::string is often a nice way to buy a bit of efficiency at the expense of a bit of safety, but the former is negligible in a method that'll only be called a handful of times at most per execution, so I'd prefer the latter.

File: src/core/src/Environment.C
Comment (roystgnr) [inline src/core/src/Environment.C:1489]: I like the name changes.  "Self-documenting code" rather than "atavistic deceptive code" is a big plus.

File: src/core/src/Environment.C
Comment (roystgnr) [inline src/core/src/Environment.C]: I'd put this comment in the header, where many people will read it as they use the API, rather than in the source file, where nobody but core developers will look unless they're already tearing their hair out over failure to understand the behavior.

File: src/core/src/EnvironmentOptions.C
Comment (roystgnr) [reply src/core/src/EnvironmentOptions.C:51]: Wait, isn't the ```new BoostInputOptionsParser()``` bit just in your patch?  The previous code was just ```m_parser()```, leaving it an uninitialized scoped pointer.

File: src/core/src/EnvironmentOptions.C
Comment (roystgnr) [reply src/core/src/EnvironmentOptions.C:169]: These parameters appear to be an atavism that haven't actually been *used* for anything:
* since 2010, at the latest
* in the version control master, rather than a branch
* in git, rather than svn

So our second best option here would be "dig through ancient svn branches to figure out what these are for", and our best would be "just get rid of them".

File: src/core/src/EnvironmentOptions.C
Comment (roystgnr) [reply src/core/src/EnvironmentOptions.C]: Nah, keep the error.  Belt-and-suspenders.

File: src/core/src/EnvironmentOptions.C
Comment (roystgnr) [reply src/core/src/EnvironmentOptions.C]: Actually, I'm pretty sure it is.  Whitespace in the input *files* is fine with either boost or getpot, but I'm not sure either trims whitespace from input *API parameters*, and if they do it's surely luck rather than officially supported behavior.

**Comment by roystgnr** [comment]
> Tests break when Boost Program Options are enabled, but they seem to on dev as well. Guidance?
> FAIL: test_BoostInputOptionsParser
> FAIL: t01_valid_cycle/rtest01.sh
> (HANGS or SLOW): test_mala

I just discovered that my gpmsa_new_functional branch doesn't even *link* when boost::program_options is enabled, so I'm merely embarrassed rather than surprised that dev isn't getting much love for that configuration either.  I could swear I had it working in January...

**Comment by roystgnr** [comment]
test_mala finished quickly for me, but the log file is filled with NaN warnings.

**Comment by roystgnr** [comment]
The other two failed.

**Comment by roystgnr** [comment]
```
  what():  the argument (''false'') for option 'bool' is invalid. Valid choices are 'on|off', 'yes|no', '1|0' and 'true|false'
```

Coincidentally, "what" was my reaction as well.

File: src/core/src/EnvironmentOptions.C
Comment (briadam) [reply src/core/src/EnvironmentOptions.C:51]: Yes, you're right.  I caused confusion here by not making reference and relaying a question I was more asking myself.  I took a cue from https://github.com/libqueso/queso/blob/c6c71a0ee2bc81c7316fe67dbff104dc8104b64e/src/gp/src/GPMSAOptions.C#L133 where it gets initialized, but I agree leaving the scoped ptr uninitialized should be okay.

File: src/core/src/EnvironmentOptions.C
Comment (briadam) [reply src/core/src/EnvironmentOptions.C:169]: Okay, if @dmcdougall agrees, I can purge them.

**Comment by roystgnr** [comment]
I saw Goody @dmcdougall [dancing with the devil](457)!

GetPot appears to be much more flexible when in how it deals with booleans and a bit trickier with strings and non-scalar options - it strips enclosing quotes if they're there, while boost::program_options gets confused by quotes around booleans and retains quotes around strings and non-scalar values.

So I think we'd actually be fine if we just used boost-compatible files for everything (and reverting the single quotes added to test01_valid_cycle/tgaCycle.inp backs me up on this)... which makes me wonder why exactly Damon went to all the trouble of changing them in the first place.

**Comment by roystgnr** [comment]
> Should I further condense code shared between FullEnvironment::construct(...) variants?

On the one hand, [I'm a big fan of refactoring that reduces code size](https://www.folklore.org/StoryView.py?story=Negative_2000_Lines_Of_Code.txt), but on the other hand only you know [if that's worth the time](https://quoteinvestigator.com/2012/04/28/shorter-letter/).

**Comment by roystgnr** [comment]
> Do we anticipate a need for options_have_been_used checking as in GPMSA?

No, but the whole point of that checking was to handle unanticipated application code mistakes, so it wouldn't hurt to use that idiom more often, if you want.

File: src/core/src/EnvironmentOptions.C
Comment (roystgnr) [reply src/core/src/EnvironmentOptions.C:51]: Ah, that looks like a mistake on my part.  If we need m_parser then we reset it anyway, so there's no point in initializing it in the first place...

**Comment by briadam** [comment]
Thanks for all the helpful review comments!
Yeah I see the same as you with t01_valid_cycle ([edit] and test_BoostInputOptionsParser).   I still have 1 fail to chase down. I'm going to focus on integrating your review first and then if you haven't picked this issue back up, can revisit.

**Comment by briadam** [comment]
> So I think we'd actually be fine if we just used boost-compatible files for everything (and reverting the single quotes added to test01_valid_cycle/tgaCycle.inp backs me up on this)... which makes me wonder why exactly Damon went to all the trouble of changing them in the first place.

I made the rest of Roy's requested changes.  Should I make a commit on this branch (or dev) to reverse the input file changes from #457 ?  Doing so seems to fix most tests with BPO enabled, though not test_intercomm0, so there might be a real regression with my changes.

**Comment by dmcdougall** [comment]
> So I think we'd actually be fine if we just used boost-compatible files for everything (and reverting the single quotes added to test01_valid_cycle/tgaCycle.inp backs me up on this)... which makes me wonder why exactly Damon went to all the trouble of changing them in the first place.

I could've sworn they weren't cross-compatible.  That's on me.

File: src/core/src/EnvironmentOptions.C
Comment (dmcdougall) [reply src/core/src/EnvironmentOptions.C:169]: I'm ok with this.

**Comment by dmcdougall** [comment]
I'm ok with doing the file-change in a different PR.

**Comment by roystgnr** [comment]
Isn't test_intercomm0 the one that we expect to randomly fail, because it doesn't use a fixed PRNG seed?  The price of stochastic regression testing...

**Comment by roystgnr** [comment]
Yeah, let's do the input format reversion in a separate PR.

**Comment by briadam** [comment]
Well good news, it (test_intercomm0) is behaving as expected then!  Issue #642 and PR #644 should resolve that part of this discussion.

**Comment by briadam** [comment]
@dmcdougall I'm working to preserve BPO support in this PR, if that affects which release we target with this.  I realize other aspects of this commit do change behavior, but at @roystgnr pointed out, only in isolated ways.  I'm thinking to take a second pass to remove BPO support once you draw a release with the first pass options refactor.

**Comment by dmcdougall** [comment]
> Isn't test_intercomm0 the one that we expect to randomly fail, because it doesn't use a fixed PRNG seed?

No.  It should pass every time.

**Comment by briadam** [comment]
@dmcdougall @roystgnr I decided to make this change in a way that's largely backward compatible and preserves Boost Program Options parsing.  Would prefer this PR make it into the next release as it's needed for effective use of GPMSA scalar capability in Dakota.

**Comment by roystgnr** [comment]
Can you fix up the conflicts?

**Comment by briadam** [comment]
Yes, working on it now.  The Github docs suggested a process that would mean I make a merge commit on dev, which is surely not right.  Should I just merge dev into my options_handling_569 branch and resolve the conflict?  Or something else?

**Comment by briadam** [comment]
Well, that'll teach me to try to do the merge with the webtool.  I'll do locally this afternoon and fix this up...

**Comment by roystgnr** [comment]
I usually "git fetch origin" and "git rebase origin/dev" and then manually fix conflicts as they come up in the rebase.  That's sometimes easier said than done, though.

**Comment by briadam** [comment]
Okay I was just being silly and had messed up that merge.  I rebased and force pushed to overwrite the bum merge.  I'm hoping this passes checks now.  

However, there's a pre-existing issue on dev that prevents building with Boost Program Options:  #655  Probably merge this PR and then see if anyone can fix that.  I struggled to get bisect working right.

**Comment by briadam** [comment]
Assuming checks pass, this can be merged.  I'll note some cleanup needs in #569 and hopefully take one more pass before calling that done.

**Comment by briadam** [comment]
I'm planing to take a cleanup pass today, including adding public default ctors to all the options classes I mucked with.

File: src/core/src/EnvironmentOptions.C
Comment (briadam) [reply src/core/src/EnvironmentOptions.C:51]: @roystgnr  FWIW, I learned the hard way why this initialization is necessary: operator<< dereferences this object, which would fail after default construction.  I added initialization to all classes for now and will promptly remove it post 0.58.0.  Same bug could affect copy ctors, but that use case isn't exercised by any code right now, so ignoring.

**Comment by briadam** [comment]
@roystgnr I rebased this on top of your GPMSAOption BPO fix.  A quick review wouldn't hurt, but it's a ton of code, and my recent revisions shouldn't be controversial.  I'm good for merge when you are.

File: src/stats/inc/StatisticalForwardProblemOptions.h
Comment (roystgnr) [inline src/stats/inc/StatisticalForwardProblemOptions.h:121]: Shoot, this fixed a memory leak, didn't it?

Thanks!

**Review by roystgnr — APPROVED** [review]

**Comment by roystgnr** [comment]
Pinging @dmcdougall - IMHO this should be good to merge, but if you've got time it wouldn't hurt to get your eyes on it too first.

**Comment by roystgnr** [comment]
@dmcdougall, in person: "I trust you."

I'm sure that'll come back to haunt him later but for now it just means we're ready to merge.

**Pull request merged**
